src/Features/Core/Shared/Utilities/PatternMatcher.cs

```diff
@@ -260,8 +288,82 @@ private static bool ContainsSpaceOrAsterisk(string text)
         /// patterns. If it was not a match, it returns null.</returns>
         public IEnumerable<PatternMatch> MatchPattern(string candidate, string pattern)
         {
+            return MatchNonDottedPattern(candidate, pattern);
+        }
+
+        /// <summary>
+        /// Matches a pattern against a candidate, and an optional dotted container for the 
+        /// candidate. If the container is provided, and the pattern itself contains dots, then
+        /// the pattern will be tested against the candidate and container.  Specifically,
+        /// the part of the pattern after the last dot will be tested against the candidate. If
+        /// a match occurs there, then the remaining dot-separated portoins of the pattern will
+        /// be tested against every successive portion of the container from right to left.
+        /// 
+        /// i.e. if you have a pattern of "Con.WL" and the candidate is "WriteLine" with a 
+        /// dotted container of "System.Console", then "WL" will be tested against "WriteLine".
+        /// With a match found there, "Con" will then be tested against "Console".
+        /// </summary>
+        public IEnumerable<PatternMatch> MatchPattern(string candidate, string dottedContainer, string pattern)
+        {
+            var dotParts = GetDotSeparatedParts(pattern);
+            return dotParts == null
+                ? MatchNonDottedPattern(candidate, pattern)
+                : MatchDottedPattern(candidate, dottedContainer, pattern, dotParts);
+        }
+
+        private IEnumerable<PatternMatch> MatchDottedPattern(string candidate, string dottedContainer, string pattern, string[] patternParts)
+        {
+            Debug.Assert(patternParts.Length > 0);
+
+            // First, check that the last part of the dot separated pattern matches the name of the
+            // candidate.  If not, then there's no point in proceeding and doing the more
+            // expensive work.
+            var candidateMatch = MatchNonDottedPattern(candidate, patternParts.Last());
+            if (candidateMatch == null)
+            {
+                return null;
+            }
+
+            // So far so good.  Now break up the container for the candidate and check if all
+            // the dotted parts match up correctly.
+            var totalMatch = candidateMatch.ToList();
```

Comment:
Do this after the check on line 335. That way, an "early out" won't incur the List<T> allocation.